jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt

```diff
@@ -0,0 +1,158 @@
+    private val acceptedTokens: AtomicInteger,
+    private val totalTokens: AtomicInteger,
+    private val rangeMarkers: MutableList<RangeMarker>
+) : Disposable {
+    val percentage: Int
+        get() = if (totalTokensSize != 0) calculatePercentage(acceptedTokensSize, totalTokensSize) else 0
+    val acceptedTokensSize: Int
+        get() = acceptedTokens.get()
+    val totalTokensSize: Int
+        get() = totalTokens.get()
+    val acceptedRecommendationsCount: Int
+        get() = rangeMarkers.size
+    private val alarm = AlarmFactory.getInstance().create(Alarm.ThreadToUse.POOLED_THREAD, this)
+    private val isShuttingDown = AtomicBoolean(false)
+    private var startTime: Instant = Instant.now()
+
+    init {
+        val conn = ApplicationManager.getApplication().messageBus.connect()
+        conn.subscribe(
+            CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
+            object : CodeWhispererUserActionListener {
+                override fun afterAccept(states: InvocationContext, sessionContext: SessionContext, rangeMarker: RangeMarker) {
+                    if (states.requestContext.fileContextInfo.programmingLanguage.toCodeWhispererLanguage() != language) return
+                    rangeMarkers.add(rangeMarker)
+                }
+            }
+        )
+        scheduleCodeWhispererCodeCoverageTracker()
+    }
+
+    fun documentChanged(event: DocumentEvent) {
+        // When open a file for the first time, IDE will also emit DocumentEvent for loading with `isWholeTextReplaced = true`
+        // Added this condition to filter out those events
+        if (event.isWholeTextReplaced) return
+        addAndGetTotalTokens(event.newLength - event.oldLength)
+    }
+
+    private fun flush() {
+        try {
+            if (isTelemetryEnabled()) emitCodeWhispererCodeContribution()
+        } finally {
+            reset()
+            scheduleCodeWhispererCodeCoverageTracker()
+        }
+    }
+
+    private fun scheduleCodeWhispererCodeCoverageTracker() {
+        if (!alarm.isDisposed && !isShuttingDown.get()) {
+            alarm.addRequest({ flush() }, Duration.ofSeconds(timeWindowInSec).toMillis())
+        }
+    }
+
+    private fun addAndGetAcceptedTokens(delta: Int): Int =
+        if (!isTelemetryEnabled()) acceptedTokensSize
+        else acceptedTokens.addAndGet(delta)
+
+    private fun addAndGetTotalTokens(delta: Int): Int =
+        if (!isTelemetryEnabled()) totalTokensSize
+        else {
+            val result = totalTokens.addAndGet(delta)
+            if (result < 0) totalTokens.set(0)
+            result
+        }
+
+    private fun reset() {
+        startTime = Instant.now()
+        totalTokens.set(0)
+        acceptedTokens.set(0)
+        rangeMarkers.clear()
+    }
+
+    private fun emitCodeWhispererCodeContribution() {
+        rangeMarkers.forEach {
+            if (!it.isValid) return@forEach
+            addAndGetAcceptedTokens(it.endOffset - it.startOffset)
+        }
```

Comment:
This is good, would be better if you can remove invalid ones in the list every 5 mins

Reply:
but we already clear all rangemarkers every 5 mins @andrewyuq , refer to line 97, is that what you mean